nmmo/task/predicate/core.py

```diff
@@ -0,0 +1,179 @@
+    return self._name
+
+  def _desc(self, class_type):
+    return {
+      "type": class_type,
+      "name": self.name,
+      "evaluate": self._evaluate.__doc__
+    }
+
+  @property
+  def description(self) -> Dict:
+    return self._desc("Predicate")
+
+  def __and__(self, other):
+    return AND(self, other)
+  def __or__(self, other):
+    return OR(self, other)
+  def __invert__(self):
+    return NOT(self)
+  def __rshift__(self, other):
+    return IMPLY(self, other)
+
+################################################
+# Sweet syntactic sugar
+def predicate(fn) -> Predicate:
+  class FunctionPredicate(Predicate):
+    def __init__(self, *args, **kwargs) -> None:
+      super().__init__(*args, name=fn.__name__, **kwargs)
+      self._args = args
+      self._kwargs = kwargs
+
+    def _evaluate(self, gs: GameState):
+      # pylint: disable=redefined-builtin, unused-variable
+      __doc__ = fn.__doc__
+      result = fn(gs, *self._args, **self._kwargs)
+      if isinstance(result, Predicate):
+        return result(gs)
+      return result
+
+  return FunctionPredicate
+
+################################################
+# Connectives
+
+class AND(Predicate):
+  def __init__(self, *predicates: Predicate):
+    super().__init__()
+    assert len(predicates) > 0
+    self._predicates = predicates
+
+    # the name is AND(task1,task2,task3)
+    self._name = 'AND(' + ','.join([t.name for t in self._predicates]) + ')'
+
+  def _evaluate(self, gs: GameState):
+    """True if all _predicates are evaluated to be True.
+    """
+    return min(t(gs) for t in self._predicates)
+
+  @property
+  def description(self) -> Dict:
+    desc = self._desc("Conjunction")
+    desc.update({ 'desc_child': ["AND"] + [t.description for t in self._predicates] })
+    return desc
+
+class OR(Predicate):
+  def __init__(self, *predicates: Predicate):
+    super().__init__()
+    assert len(predicates) > 0
+    self._predicates = predicates
+
+    # the name is OR(task1,task2,task3,...)
+    self._name = 'OR(' + ','.join([t.name for t in self._predicates]) + ')'
+
+  def _evaluate(self, gs: GameState):
+    """True if any of _predicates is evaluated to be True.
+    """
+    return max(t(gs) for t in self._predicates)
+
+  @property
+  def description(self) -> Dict:
+    desc = self._desc("Disjunction")
+    desc.update({ 'desc_child': ["OR"] + [t.description for t in self._predicates] })
+    return desc
+
+class NOT(Predicate):
+  def __init__(self, task: Predicate):
+    super().__init__()
+    # pylint: disable=super-init-not-called
+    self._task = task
+
+    # the name is NOT(task)
+    self._name = f'NOT({self._task.name})'
+
+  def _evaluate(self, gs: GameState):
+    """True if _task is evaluated to be False.
+    """
+    return 1- self._task(gs)
+
+  @property
+  def description(self) -> Dict:
+    desc = self._desc("Negation")
+    desc.update({ 'desc_child': ["NOT", self._task.description] })
+    return desc
+
+class IMPLY(Predicate):
```

Comment:
Do we really want this predicate? I'm having a hard time thinking of a use case.

Reply:
Follows from first-order logic. Technically we'd want EXISTS and FORALL too then.... but these wrappers are peripheral anyways